proxy/proxy.go

```diff
@@ -467,24 +470,60 @@ func setRequestURLForDynamicBackend(u *url.URL, stateBag map[string]interface{})
 	}
 }
 
-func selectEndpoint(ctx *context, registry *routing.EndpointRegistry) *routing.LBEndpoint {
+func fadeIn(now time.Time, duration time.Duration, exponent float64, detected time.Time) float64 {
+	rel := now.Sub(detected)
+	fadingIn := rel > 0 && rel < duration
+	if !fadingIn {
+		return 1
+	}
+
+	return math.Pow(float64(rel)/float64(duration), exponent)
+}
+
+func filterFadeIn(endpoints []routing.LBEndpoint, rt *routing.Route, registry *routing.EndpointRegistry, rnd *rand.Rand) []routing.LBEndpoint {
+	now := time.Now()
+	threshold := rnd.Float64()
+
+	filtered := make([]routing.LBEndpoint, 0, len(endpoints))
+	for _, e := range endpoints {
+		detected := registry.GetMetrics(e.Host).DetectedTime()
+		f := fadeIn(
+			now,
+			rt.LBFadeInDuration,
+			rt.LBFadeInExponent,
+			detected,
+		)
+		if threshold < f {
+			filtered = append(filtered, e)
+		}
+	}
+	return filtered
+}
+
+func selectEndpoint(ctx *context, registry *routing.EndpointRegistry, rnd *rand.Rand) *routing.LBEndpoint {
 	rt := ctx.route
+	endpoints := rt.LBEndpoints
+	endpoints = filterFadeIn(endpoints, rt, registry, rnd)
+	if len(endpoints) == 0 {
+		endpoints = rt.LBEndpoints
+	}
```

Comment:
Maybe this should be moved into filterFadeIn().
The idea is that we have a series of endpoint filters here like:
```
endpoints := rt.LBEndpoints
endpoints = filterFoo(endpoints)
endpoints = filterBar(endpoints)
endpoints = filterBaz(endpoints)
```
and if some filter sees that it produces no results it returns its input unmodified such that result of previous filters is not lost (which would be the case if we reset to `endpoints = rt.LBEndpoints`).